src/ucp/wireup/wireup_cm.c

```diff
@@ -41,6 +41,73 @@ int ucp_ep_init_flags_has_cm(unsigned ep_init_flags)
                                UCP_EP_INIT_CM_WIREUP_SERVER));
 }
 
+static const char* ucp_context_get_cm_name_by_index(ucp_context_h context,
+                                                    ucp_rsc_index_t cm_idx,
+                                                    char *name, size_t max_length)
+{
+    ucp_rsc_index_t cmpt_index = context->config.cm_cmpt_idxs[cm_idx];
+
+    ucs_snprintf_safe(name, max_length, "%s", context->tl_cmpts[cmpt_index].attr.name);
+    return name;
+}
+
+/*
+ * The main thread progress part of attempting connecting the client to the server
+ * through the next available cm.
+ */
+static unsigned ucp_cm_client_try_next_cm_progress(void *arg)
+{
+    ucp_ep_h ucp_ep       = arg;
+    ucp_worker_h worker   = ucp_ep->worker;
+    ucp_context_h context = worker->context;
+    ucp_wireup_ep_t *cm_wireup_ep;
+    ucs_status_t status;
+    char cm_name_prev[UCT_TL_NAME_MAX], cm_name_next[UCT_TL_NAME_MAX];
+
+    UCS_ASYNC_BLOCK(&worker->async);
+
+    cm_wireup_ep = ucp_ep_get_cm_wireup_ep(ucp_ep);
+    ucs_assert_always(cm_wireup_ep != NULL);
+    ucp_wireup_ep_destroy_next_ep(cm_wireup_ep);
+
+    /* try connecting through another cm (next one in the priority list) */
+    cm_wireup_ep->cm_idx++;
+
+    ucs_debug("client switching from %s to %s in attempt to connect to the server",
+              ucp_context_get_cm_name_by_index(context, cm_wireup_ep->cm_idx - 1,
+                                               cm_name_prev, UCT_TL_NAME_MAX),
+              ucp_context_get_cm_name_by_index(context, cm_wireup_ep->cm_idx,
+                                               cm_name_next, UCT_TL_NAME_MAX));
+
+    status = ucp_ep_client_cm_create_uct_ep(ucp_ep);
+    if (status != UCS_OK) {
+        ucs_error("failed to create a uct sockaddr endpoint on %s cm %p",
+                  ucp_context_get_cm_name_by_index(context, cm_wireup_ep->cm_idx,
+                                                   cm_name_next, UCT_TL_NAME_MAX),
+                  worker->cms[cm_wireup_ep->cm_idx].cm);
+
+        uct_ep_destroy(&cm_wireup_ep->super.super);
+        ucp_ep_destroy_internal(ucp_ep);
+
+        UCS_ASYNC_UNBLOCK(&worker->async);
+        return 0;
+    }
+
+    UCS_ASYNC_UNBLOCK(&worker->async);
+    return 1;
+}
+
+static void ucp_cm_client_fallback_cms(ucp_worker_h worker, ucp_ep_h ep)
```

Comment:
rename ucp_cm_client_fallback_cms -> ucp_cm_client_try_next_cm
increment and check index here before progress func register and make sure if it's single place

Reply:
why do the increment here and not right before it's used?

Reply:
to move related code closer